File: year2025/day5/part1/solution.py
#!/usr/bin/env python3
# Day 5: Cafeteria, Part 1

FILE_PATH = "../.input"

def main():
    total_fresh = 0
    id_ranges = []
    ingredient_ids = []

    with open(FILE_PATH, 'r') as ids_file:
        range_flag = True  # Was debating using split here on every line and checking length of results, but this will probably translate better anyways? May also be more efficient
        for line in ids_file:
            line = line.strip()
            if line == '':
                range_flag = False
                continue
            if range_flag:
                id_ranges.append(tuple(map(int, line.split('-'))))
            else:
                ingredient_ids.append(int(line))  # Could technically make this in place rather that storing in memory

    for ingredient_id in ingredient_ids:
        for bot, top in id_ranges:
            # print(f"DEBUG: {bot}, {ingredient_id}, {top}")
            if bot <= ingredient_id <= top:
                total_fresh += 1
                break

    # print(f"DEBUG: {id_ranges}")
    # print(f"DEBUG: {ingredient_ids}")
    print(f"Total Fresh IDs: {total_fresh}")

if __name__ == "__main__":
    main()

File: year2025/day5/part1/solution.cpp
// Day 5: Cafeteria, Part 1

#include <iostream>
#include <fstream>
#include <string>
#include <vector>

using namespace std;

const string FILE_PATH = "../.input";

int main() {
    unsigned long long total_fresh = 0;
    vector<pair<unsigned long long, unsigned long long>> id_ranges;
    vector<unsigned long long> ingredient_ids;

    ifstream ids_file(FILE_PATH);
    string line;
    bool range_flag = true;

    while(getline(ids_file, line)) {
        if (line == "") {
            range_flag = false;
            continue;
        }
        if (range_flag) {
            size_t dash_pos = line.find('-');
            string first_str = line.substr(0, dash_pos);
            string second_str = line.substr(dash_pos + 1);
            //cout << "DEBUG: first: " << first_str << ", second: " << second_str << endl;

            unsigned long long first_num = stoull(first_str);
            unsigned long long second_num = stoull(second_str);

            id_ranges.emplace_back(make_pair(first_num, second_num));
        } else {
            ingredient_ids.emplace_back(stoull(line));
        }
    }

    for (const unsigned long long ingredient_id : ingredient_ids) {
        for (const pair<unsigned long long, unsigned long long>& range : id_ranges) {
            //cout << "DEBUG: " << range.first << ", " << ingredient_id << ", " << range.second << endl;
            if ((range.first <= ingredient_id) && (ingredient_id <= range.second)) {
                ++total_fresh;
                break;
            }
        }
    }

    //for (const pair<unsigned long long, unsigned long long>& range : id_ranges) {
    //    cout << "DEBUG: Range: " << range.first << "-" << range.second << "\n";
    //}
    //for (const unsigned long long ingredient_id : ingredient_ids) {
    //    cout << "DEBUG: ID: " << ingredient_id << "\n";
    //}
    cout << "Total Fresh IDs: " << total_fresh << endl;

    return 0;
}
